java/ql/src/experimental/Security/CWE/CWE-094/TemplateInjection.qll

```diff
@@ -0,0 +1,241 @@
+}
+
+/**
+ * An argument to MVEL template engine's `compileTemplate` method call is as a sink for `Configuration`.
+ */
+private class MVELCompileTemplateSink extends Sink {
+  MVELCompileTemplateSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getArgument(0) and
+      ma.getMethod() instanceof MethodMVELCompileTemplate
+    )
+  }
+}
+
+/**
+ * An argument to MVEL templating engine's `TemplateCompiler`
+ *   construtor call is a sink for `Configuration`.
+ */
+private class MVELConstructorSink extends Sink {
+  MVELConstructorSink() {
+    exists(ConstructorCall cc, Constructor c |
+      cc.getConstructor() = c and
+      c.getDeclaringType() instanceof MVELTemplateCompiler and
+      this.asExpr() = cc.getArgument(0)
+    )
+  }
+}
+
+/**
+ * Tainted data flowing into a Velocity Context through `put` method taints the context.
+ */
+private class VelocityContextFlow extends AdditionalFlowStep {
+  override predicate isAdditionalTaintStep(DataFlow::Node prev, DataFlow::Node succ) {
+    exists(MethodAccess m | m.getMethod() instanceof MethodVelocityContextPut |
+      m.getArgument(1) = prev.asExpr() and
+      (succ.asExpr() = m or succ.asExpr() = m.getQualifier())
+    )
+  }
+}
+
+/**
+ * An argument to Velocity template engine's `mergeTemplate` method call is a sink for `Configuration`.
+ */
+private class VelocityMergeTempSink extends Sink {
+  VelocityMergeTempSink() {
+    exists(MethodAccess m |
+      // static boolean	mergeTemplate(String templateName, String encoding, Context context, Writer writer)
+      m.getCallee() instanceof MethodVelocityMergeTemplate and
+      m.getArgument(2) = this.getExpr()
+    )
+  }
+}
+
+/**
+ * An argument to Velocity template engine's `evaluate` method call is a sink for `Configuration`.
+ */
+private class VelocityEvaluateSink extends Sink {
+  VelocityEvaluateSink() {
+    exists(MethodAccess m |
+      m.getCallee() instanceof MethodVelocityEvaluate and
+      m.getArgument([0, 3]) = this.getExpr()
+    )
+  }
+}
+
+/**
+ * An argument to Velocity template engine's `parse` method call is a sink for `Configuration`.
+ */
+private class VelocityParseSink extends Sink {
+  VelocityParseSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getArgument(0) and
+      ma.getMethod() instanceof MethodVelocityParse
+    )
+  }
+}
+
+/**
+ * An argument to Velocity template engine's `putStringResource` method call is a sink for `Configuration`.
+ */
+private class VelocityPutStringResSink extends DataFlow::Node {
+  VelocityPutStringResSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getArgument(1) and
+      ma.getMethod() instanceof MethodVelocityPutStringResource
+    )
+  }
+}
+
+/**
+ * An argument to Velocity template engine's `addVelocimacro` method call is a sink for `Configuration`.
+ */
+private class VelocityAddVelociMacroSink extends Sink {
+  VelocityAddVelociMacroSink() {
+    exists(MethodAccess ma |
+      this.asExpr() = ma.getArgument(1) and
```

Comment:
How is a [`Node`](https://velocity.apache.org/engine/2.3/apidocs/org/apache/velocity/runtime/RuntimeSingleton.html#addVelocimacro-java.lang.String-org.apache.velocity.runtime.parser.node.Node-java.util.List-org.apache.velocity.Template-) supposed to be tainted? I can't see any taint steps modeling it.

Reply:
@atorralba It was meant to model [this](https://velocity.apache.org/engine//1.6.2/apidocs/org/apache/velocity/runtime/VelocimacroFactory.html#addVelocimacro(java.lang.String,%20java.lang.String,%20java.lang.String[],%20java.lang.String)) method. 

The API has since changed. Now that I think about this, we shouldn't mark it as a sink at all. The API is too low level for QL to concern with. A [quick search ](https://grep.app/search?q=addVelocimacro)on grep.app reveals only 13 results. And even then, the usage is only internal.

Hence, I have decided to drop this as a sink